impl-tinkerpop/src/main/java/org/hawkular/inventory/impl/tinkerpop/AbstractSourcedGraphService.java

```diff
@@ -104,4 +234,90 @@ protected void removeRelationship(Constants.Type typeInSource, Relationships.Wel
     protected abstract String getProposedId(Blueprint b);
 
     protected abstract Filter[] initNewEntity(Vertex newEntity, Blueprint blueprint);
+
+    private static class BFSTransitiveClosureSpliterator implements Spliterator<Vertex> {
+        private final Direction direction;
+        private final String[] edgeLabels;
+        Deque<Iterator<Vertex>> pending;
+
+        private BFSTransitiveClosureSpliterator(Iterator<Vertex> pending, Direction direction,
+                                                String[] edgeLabels) {
+            this.direction = direction;
+            this.edgeLabels = edgeLabels;
+            this.pending = new ArrayDeque<>(Arrays.asList(pending));
+        }
+
+        @Override
+        public boolean tryAdvance(Consumer<? super Vertex> action) {
+            Iterator<Vertex> queueToProcess = peekNonEmpty();
+            if (queueToProcess == null) {
+                return false;
+            }
+
+            Vertex toProcess = queueToProcess.next();
+
+            pending.add(toProcess.getVertices(direction, edgeLabels).iterator());
+
+            action.accept(toProcess);
```

Comment:
I just tried and it's not working on Titan (and because my Stream uses the same approach I assume it wouldn't work, too).

In your example graph above, running:
```java
v1.out("contains").loop(1){true}{true}.remove()
```

will barf with "unable to access a removed vertex" error. That is because the loop (as well as my stream) will return vertices as soon as it finds them - and the pipe will call `remove()` on them as they flow in. So you first remove a vertex and then want to find other vertices by looking at the "contains" edges of the element you just removed.

This needs to go "bottom up" instead. Any ideas for some Gremlin magic to do that?